File: ssh-manager/ssh_manager.py
#!/usr/bin/env python3
"""
jump — Personal SSH connection manager
Usage: python ssh_manager.py [command] [args]
       python ssh_manager.py          (interactive menu)
"""

import os
import sys
import json
import subprocess
import socket
import argparse
import shutil
from pathlib import Path
from datetime import datetime

CONFIG_DIR = Path.home() / ".ssh_manager"
DEVICES_FILE = CONFIG_DIR / "devices.json"
APP_NAME = "jump"
VERSION = "1.0.0"

# ── ANSI colors (works on Windows 10+ with VT processing enabled) ──────────
if sys.platform == "win32":
    os.system("")  # enable VT100 on Windows console

R = "\033[0m"
BOLD = "\033[1m"
DIM = "\033[2m"
RED = "\033[91m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
BLUE = "\033[94m"
MAGENTA = "\033[95m"
CYAN = "\033[96m"
WHITE = "\033[97m"

def c(color, text):
    return f"{color}{text}{R}"

# ── Storage ─────────────────────────────────────────────────────────────────
def load_devices():
    if DEVICES_FILE.exists():
        with open(DEVICES_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    return {}

def save_devices(devices):
    CONFIG_DIR.mkdir(parents=True, exist_ok=True)
    with open(DEVICES_FILE, "w", encoding="utf-8") as f:
        json.dump(devices, f, indent=2)

# ── Formatting helpers ───────────────────────────────────────────────────────
def banner():
    print(f"""
{BOLD}{CYAN}  ╔══════════════════════════════╗
  ║  ✈  jump  —  ssh manager   ║
  ╚══════════════════════════════╝{R}  {DIM}v{VERSION}{R}
""")

def divider(label=""):
    width = 54
    if label:
        pad = (width - len(label) - 2) // 2
        print(f"{DIM}{'─' * pad} {label} {'─' * (width - pad - len(label) - 2)}{R}")
    else:
        print(f"{DIM}{'─' * width}{R}")

def ok(msg):   print(f"  {GREEN}✓{R}  {msg}")
def warn(msg): print(f"  {YELLOW}⚠{R}  {msg}")
def err(msg):  print(f"  {RED}✗{R}  {msg}")
def info(msg): print(f"  {CYAN}→{R}  {msg}")

def prompt(label, default=None, secret=False):
    hint = f" [{DIM}{default}{R}]" if default else ""
    try:
        if secret:
            import getpass
            val = getpass.getpass(f"  {BOLD}{label}{hint}: ")
        else:
            val = input(f"  {BOLD}{label}{hint}: ").strip()
        return val if val else default
    except (KeyboardInterrupt, EOFError):
        print()
        return default

def device_tag(d):
    tag = f"{d.get('user', '')}@{d['host']}"
    if d.get('port', 22) != 22:
        tag += f":{d['port']}"
    return tag

# ── Device table ─────────────────────────────────────────────────────────────
def print_table(devices):
    if not devices:
        warn("No devices saved yet. Run  add  to add one.")
        return

    col_name  = max(len("NAME"),  max(len(n) for n in devices))
    col_host  = max(len("HOST/IP"), max(len(d['host']) for d in devices.values()))
    col_user  = max(len("USER"),  max(len(d.get('user','')) for d in devices.values()))
    col_port  = 5
    col_group = max(len("GROUP"), max(len(d.get('group','—')) for d in devices.values()))

    hdr = (f"  {BOLD}{CYAN}{'NAME':<{col_name}}{R}  "
           f"{BOLD}{'HOST/IP':<{col_host}}{R}  "
           f"{BOLD}{'USER':<{col_user}}{R}  "
           f"{BOLD}{'PORT':<{col_port}}{R}  "
           f"{BOLD}{'GROUP':<{col_group}}{R}  "
           f"{BOLD}NOTES{R}")
    print(hdr)
    divider()

    for name, d in sorted(devices.items()):
        online_sym = ""
        port   = d.get('port', 22)
        group  = d.get('group', '—')
        notes  = d.get('notes', '')
        user   = d.get('user', '')
        host_str = c(YELLOW, f"{d['host']:<{col_host}}")
        port_str = c(DIM, str(port))
        print(f"  {c(BOLD+GREEN, f'{name:<{col_name}}')}  "
              f"{host_str}  "
              f"{user:<{col_user}}  "
              f"{port_str:<{col_port + len(DIM) + len(R)}}  "
              f"{group:<{col_group}}  "
              f"{c(DIM, notes)}")
    divider()
    print(f"  {DIM}{len(devices)} device(s){R}\n")

# ── Commands ─────────────────────────────────────────────────────────────────
def cmd_list(devices, _args=None):
    banner()
    divider("saved devices")
    print_table(devices)

def cmd_add(devices, args=None):
    print(f"\n{BOLD}  Add a new device{R}\n")
    # Name
    name = None
    while not name:
        name = prompt("Alias (short name)", "").strip()
        if not name:
            err("Alias cannot be empty.")
        elif name in devices:
            err(f"'{name}' already exists. Use  edit  to update it.")
            name = None

    host    = prompt("Hostname or IP")
    user    = prompt("SSH username", os.environ.get("USERNAME") or os.environ.get("USER") or "")
    port    = prompt("Port", "22")
    key     = prompt("Identity file (path to .pem/.key, leave blank for default)", "")
    group   = prompt("Group / tag", "")
    notes   = prompt("Notes", "")

    try:
        port = int(port)
    except (ValueError, TypeError):
        port = 22

    device = {
        "host":    host,
        "user":    user,
        "port":    port,
        "added":   datetime.now().isoformat(timespec="seconds"),
    }
    if key:    device["identity"] = key
    if group:  device["group"]    = group
    if notes:  device["notes"]    = notes

    devices[name] = device
    save_devices(devices)
    ok(f"Saved  {c(BOLD, name)}  →  {device_tag(device)}\n")

def cmd_connect(devices, args):
    name = args.name if hasattr(args, 'name') else args
    if name not in devices:
        err(f"Unknown device '{name}'. Run  list  to see saved devices.")
        return
    d = devices[name]
    _connect(name, d)

def _connect(name, d):
    host  = d["host"]
    user  = d.get("user", "")
    port  = d.get("port", 22)
    key   = d.get("identity", "")

    cmd = ["ssh"]
    if port != 22:
        cmd += ["-p", str(port)]
    if key:
        cmd += ["-i", key]
    if user:
        cmd.append(f"{user}@{host}")
    else:
        cmd.append(host)

    info(f"Connecting to {c(BOLD, name)}  ({' '.join(cmd)})\n")

    # Update last_used
    devices_live = load_devices()
    if name in devices_live:
        devices_live[name]["last_used"] = datetime.now().isoformat(timespec="seconds")
        save_devices(devices_live)

    try:
        subprocess.run(cmd)
    except FileNotFoundError:
        err("'ssh' not found. Make sure OpenSSH is installed and in your PATH.")
    except KeyboardInterrupt:
        print()
        info("Session ended.")

def cmd_remove(devices, args):
    name = args.name if hasattr(args, 'name') else args
    if name not in devices:
        err(f"No device named '{name}'.")
        return
    confirm = prompt(f"Delete '{name}'? (yes/no)", "no")
    if confirm.lower() in ("y", "yes"):
        del devices[name]
        save_devices(devices)
        ok(f"Removed '{name}'.")
    else:
        info("Cancelled.")

def cmd_edit(devices, args):
    name = args.name if hasattr(args, 'name') else args
    if name not in devices:
        err(f"No device named '{name}'.")
        return
    d = devices[name]
    print(f"\n{BOLD}  Editing '{name}'  {DIM}(press Enter to keep current value){R}\n")
    d["host"]  = prompt("Hostname or IP",    d.get("host", "")) or d["host"]
    d["user"]  = prompt("Username",          d.get("user", "")) or d.get("user", "")
    d["port"]  = int(prompt("Port",          str(d.get("port", 22))) or 22)
    d["identity"] = prompt("Identity file",  d.get("identity", ""))
    d["group"]    = prompt("Group",          d.get("group", ""))
    d["notes"]    = prompt("Notes",          d.get("notes", ""))
    # strip empty optional fields
    for k in ("identity", "group", "notes"):
        if not d.get(k):
            d.pop(k, None)
    devices[name] = d
    save_devices(devices)
    ok(f"Updated '{name}'.\n")

def cmd_ping(devices, args):
    name = args.name if hasattr(args, 'name') else args
    if name not in devices:
        err(f"No device named '{name}'.")
        return
    d = devices[name]
    host = d["host"]
    port = d.get("port", 22)
    info(f"Checking {host}:{port} …")
    try:
        with socket.create_connection((host, port), timeout=5):
            ok(f"{host}:{port} is {c(GREEN, 'reachable')} (SSH port open)\n")
    except (socket.timeout, ConnectionRefusedError, OSError) as e:
        err(f"{host}:{port} is {c(RED, 'unreachable')} — {e}\n")

def cmd_copy_id(devices, args):
    name = args.name if hasattr(args, 'name') else args
    if name not in devices:
        err(f"No device named '{name}'.")
        return
    d = devices[name]
    pub_key = prompt("Path to public key", str(Path.home() / ".ssh" / "id_rsa.pub"))
    cmd = ["ssh-copy-id"]
    port = d.get("port", 22)
    if port != 22:
        cmd += ["-p", str(port)]
    if d.get("identity"):
        cmd += ["-i", d["identity"]]
    user = d.get("user", "")
    host = d["host"]
    cmd.append(f"{user}@{host}" if user else host)
    info(f"Running: {' '.join(cmd)}")
    try:
        subprocess.run(cmd)
    except FileNotFoundError:
        err("'ssh-copy-id' not found. On Windows, copy the key manually.")

def cmd_export(devices, args=None):
    out = Path(prompt("Export to file", "jump_devices.json"))
    with open(out, "w", encoding="utf-8") as f:
        json.dump(devices, f, indent=2)
    ok(f"Exported {len(devices)} device(s) to {out}\n")

def cmd_import(devices, args=None):
    src = Path(prompt("Import from file", "jump_devices.json"))
    if not src.exists():
        err(f"File not found: {src}")
        return
    with open(src, "r", encoding="utf-8") as f:
        incoming = json.load(f)
    merged = 0
    for name, d in incoming.items():
        if name in devices:
            overwrite = prompt(f"'{name}' exists — overwrite? (yes/no)", "no")
            if overwrite.lower() not in ("y", "yes"):
                continue
        devices[name] = d
        merged += 1
    save_devices(devices)
    ok(f"Imported {merged} device(s).\n")

# ── Interactive menu ──────────────────────────────────────────────────────────
def interactive_menu():
    devices = load_devices()
    banner()
    while True:
        divider("saved devices")
        print_table(devices)
        divider("actions")
        actions = [
            ("c", "connect",     "SSH into a device"),
            ("a", "add",         "Add new device"),
            ("e", "edit",        "Edit a device"),
            ("r", "remove",      "Remove a device"),
            ("p", "ping",        "Ping (check reachability)"),
            ("k", "copy-id",     "Copy SSH key to device"),
            ("x", "export",      "Export devices to JSON"),
            ("i", "import",      "Import devices from JSON"),
            ("q", "quit",        "Exit"),
        ]
        for key, name, desc in actions:
            print(f"  {c(BOLD+CYAN, f'[{key}]')}  {c(BOLD, name):<10}  {c(DIM, desc)}")
        print()

        try:
            choice = input(f"  {BOLD}>{R} ").strip().lower()
        except (KeyboardInterrupt, EOFError):
            print()
            break

        print()

        if choice in ("q", "quit", "exit"):
            info("Bye!\n")
            break

        elif choice in ("c", "connect"):
            if not devices:
                warn("No devices. Add one first.")
                continue
            names = sorted(devices)
            for i, n in enumerate(names, 1):
                print(f"  {c(CYAN, str(i)+'.')} {n}  {c(DIM, device_tag(devices[n]))}")
            print()
            sel = prompt("Device name or number")
            if not sel:
                continue
            if sel.isdigit():
                idx = int(sel) - 1
                if 0 <= idx < len(names):
                    sel = names[idx]
                else:
                    err("Invalid number.")
                    continue
            if sel in devices:
                _connect(sel, devices[sel])
                devices = load_devices()  # refresh after session
            else:
                err(f"No device '{sel}'.")

        elif choice in ("a", "add"):
            cmd_add(devices)

        elif choice in ("e", "edit"):
            name = prompt("Device name to edit")
            if name:
                cmd_edit(devices, name)
                devices = load_devices()

        elif choice in ("r", "remove"):
            name = prompt("Device name to remove")
            if name:
                cmd_remove(devices, name)
                devices = load_devices()

        elif choice in ("p", "ping"):
            name = prompt("Device name to ping")
            if name:
                cmd_ping(devices, name)

        elif choice in ("k", "copy-id"):
            name = prompt("Device name")
            if name:
                cmd_copy_id(devices, name)

        elif choice in ("x", "export"):
            cmd_export(devices)

        elif choice in ("i", "import"):
            cmd_import(devices)
            devices = load_devices()

        else:
            warn(f"Unknown command '{choice}'.")

# ── CLI entry point ───────────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(
        prog=APP_NAME,
        description="jump — personal SSH connection manager",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
commands:
  list                     list all saved devices
  add                      add a new device (interactive)
  connect <name>           SSH into a device
  remove  <name>           delete a saved device
  edit    <name>           edit a device's settings
  ping    <name>           check if SSH port is reachable
  copy-id <name>           copy your public key to a device
  export                   export devices to JSON
  import                   import devices from JSON

  (no command)             launch interactive menu
        """,
    )
    parser.add_argument("--version", action="version", version=f"{APP_NAME} {VERSION}")
    sub = parser.add_subparsers(dest="command")

    sub.add_parser("list")
    sub.add_parser("add")

    for cmd in ("connect", "remove", "edit", "ping", "copy-id"):
        p = sub.add_parser(cmd)
        p.add_argument("name")

    sub.add_parser("export")
    sub.add_parser("import")

    args = parser.parse_args()
    devices = load_devices()

    if args.command is None:
        interactive_menu()
    elif args.command == "list":
        cmd_list(devices)
    elif args.command == "add":
        cmd_add(devices)
    elif args.command == "connect":
        cmd_connect(devices, args)
    elif args.command == "remove":
        cmd_remove(devices, args)
    elif args.command == "edit":
        cmd_edit(devices, args)
    elif args.command == "ping":
        cmd_ping(devices, args)
    elif args.command == "copy-id":
        cmd_copy_id(devices, args)
    elif args.command == "export":
        cmd_export(devices)
    elif args.command == "import":
        cmd_import(devices)


if __name__ == "__main__":
    main()

File: ssh-manager/ssh_manager.js
#!/usr/bin/env node
/**
 * jump — Personal SSH connection manager
 * Usage:  node ssh_manager.js [command] [args]
 *         node ssh_manager.js          (interactive menu)
 */

"use strict";

const os   = require("os");
const fs   = require("fs");
const path = require("path");
const net  = require("net");
const { spawnSync, execSync } = require("child_process");
const readline = require("readline");

const CONFIG_DIR   = path.join(os.homedir(), ".ssh_manager");
const DEVICES_FILE = path.join(CONFIG_DIR, "devices.json");
const VERSION      = "1.0.0";

// ── ANSI colors ──────────────────────────────────────────────────────────────
const R      = "\x1b[0m";
const BOLD   = "\x1b[1m";
const DIM    = "\x1b[2m";
const RED    = "\x1b[91m";
const GREEN  = "\x1b[92m";
const YELLOW = "\x1b[93m";
const BLUE   = "\x1b[94m";
const CYAN   = "\x1b[96m";
const WHITE  = "\x1b[97m";

const c = (color, text) => `${color}${text}${R}`;

// ── Storage ──────────────────────────────────────────────────────────────────
function loadDevices() {
  if (fs.existsSync(DEVICES_FILE)) {
    return JSON.parse(fs.readFileSync(DEVICES_FILE, "utf8"));
  }
  return {};
}

function saveDevices(devices) {
  fs.mkdirSync(CONFIG_DIR, { recursive: true });
  fs.writeFileSync(DEVICES_FILE, JSON.stringify(devices, null, 2), "utf8");
}

// ── Print helpers ─────────────────────────────────────────────────────────────
function banner() {
  console.log(`
${BOLD}${CYAN}  ╔══════════════════════════════╗
  ║  ✈  jump  —  ssh manager   ║
  ╚══════════════════════════════╝${R}  ${DIM}v${VERSION}${R}
`);
}

function divider(label = "") {
  const width = 54;
  if (label) {
    const pad = Math.floor((width - label.length - 2) / 2);
    const right = width - pad - label.length - 2;
    console.log(`${DIM}${"─".repeat(pad)} ${label} ${"─".repeat(right)}${R}`);
  } else {
    console.log(`${DIM}${"─".repeat(width)}${R}`);
  }
}

const ok   = (m) => console.log(`  ${GREEN}✓${R}  ${m}`);
const warn = (m) => console.log(`  ${YELLOW}⚠${R}  ${m}`);
const err  = (m) => console.log(`  ${RED}✗${R}  ${m}`);
const info = (m) => console.log(`  ${CYAN}→${R}  ${m}`);

function deviceTag(d) {
  let tag = `${d.user || ""}@${d.host}`;
  if ((d.port || 22) !== 22) tag += `:${d.port}`;
  return tag;
}

function pad(str, len) {
  const visible = str.replace(/\x1b\[[0-9;]*m/g, "");
  return str + " ".repeat(Math.max(0, len - visible.length));
}

// ── Device table ──────────────────────────────────────────────────────────────
function printTable(devices) {
  const names = Object.keys(devices);
  if (names.length === 0) {
    warn("No devices saved yet. Run  add  to add one.");
    return;
  }

  const colName  = Math.max(4, ...names.map(n => n.length));
  const colHost  = Math.max(7, ...names.map(n => (devices[n].host || "").length));
  const colUser  = Math.max(4, ...names.map(n => (devices[n].user || "").length));
  const colPort  = 5;
  const colGroup = Math.max(5, ...names.map(n => (devices[n].group || "—").length));

  const hdr = [
    `  ${BOLD}${CYAN}${"NAME".padEnd(colName)}${R}`,
    `${BOLD}${"HOST/IP".padEnd(colHost)}${R}`,
    `${BOLD}${"USER".padEnd(colUser)}${R}`,
    `${BOLD}${"PORT".padEnd(colPort)}${R}`,
    `${BOLD}${"GROUP".padEnd(colGroup)}${R}`,
    `${BOLD}NOTES${R}`,
  ].join("  ");
  console.log(hdr);
  divider();

  for (const name of names.sort()) {
    const d     = devices[name];
    const group = d.group || "—";
    const notes = d.notes || "";
    const user  = d.user  || "";
    const port  = d.port  || 22;

    const row = [
      `  ${pad(c(BOLD + GREEN, name), colName + BOLD.length + GREEN.length + R.length)}`,
      pad(c(YELLOW, d.host), colHost + YELLOW.length + R.length),
      user.padEnd(colUser),
      pad(c(DIM, String(port)), colPort + DIM.length + R.length),
      group.padEnd(colGroup),
      c(DIM, notes),
    ].join("  ");
    console.log(row);
  }

  divider();
  console.log(`  ${DIM}${names.length} device(s)${R}\n`);
}

// ── Prompt (async readline) ───────────────────────────────────────────────────
function askQuestion(rl, label, defaultVal = "") {
  const hint = defaultVal ? ` [${DIM}${defaultVal}${R}]` : "";
  return new Promise((resolve) => {
    rl.question(`  ${BOLD}${label}${R}${hint}: `, (ans) => {
      resolve(ans.trim() || defaultVal);
    });
  });
}

// ── Commands ──────────────────────────────────────────────────────────────────
function cmdList(devices) {
  banner();
  divider("saved devices");
  printTable(devices);
}

async function cmdAdd(devices, rl) {
  console.log(`\n${BOLD}  Add a new device${R}\n`);

  let name;
  while (!name) {
    name = await askQuestion(rl, "Alias (short name)");
    if (!name) {
      err("Alias cannot be empty.");
      name = null;
    } else if (devices[name]) {
      err(`'${name}' already exists. Use  edit  to update it.`);
      name = null;
    }
  }

  const host  = await askQuestion(rl, "Hostname or IP");
  const user  = await askQuestion(rl, "SSH username", os.userInfo().username || "");
  const portS = await askQuestion(rl, "Port", "22");
  const key   = await askQuestion(rl, "Identity file (.pem/.key path, blank = default)", "");
  const group = await askQuestion(rl, "Group / tag", "");
  const notes = await askQuestion(rl, "Notes", "");

  const port = parseInt(portS, 10) || 22;

  const device = {
    host,
    user,
    port,
    added: new Date().toISOString(),
  };
  if (key)   device.identity = key;
  if (group) device.group    = group;
  if (notes) device.notes    = notes;

  devices[name] = device;
  saveDevices(devices);
  ok(`Saved  ${c(BOLD, name)}  →  ${deviceTag(device)}\n`);
}

function doConnect(name, d) {
  const host = d.host;
  const user = d.user || "";
  const port = d.port || 22;
  const key  = d.identity || "";

  const args = [];
  if (port !== 22) args.push("-p", String(port));
  if (key)         args.push("-i", key);
  args.push(user ? `${user}@${host}` : host);

  info(`Connecting to ${c(BOLD, name)}  (ssh ${args.join(" ")})\n`);

  // Update last_used
  const live = loadDevices();
  if (live[name]) {
    live[name].last_used = new Date().toISOString();
    saveDevices(live);
  }

  const result = spawnSync("ssh", args, { stdio: "inherit", shell: process.platform === "win32" });
  if (result.error && result.error.code === "ENOENT") {
    err("'ssh' not found. Make sure OpenSSH is installed and in your PATH.");
    info("On Windows: Settings → Apps → Optional Features → OpenSSH Client");
  }
}

async function cmdEdit(devices, name, rl) {
  if (!devices[name]) { err(`No device named '${name}'.`); return; }
  const d = devices[name];
  console.log(`\n${BOLD}  Editing '${name}'  ${DIM}(Enter to keep current)${R}\n`);

  d.host     = await askQuestion(rl, "Hostname or IP",        d.host     || "");
  d.user     = await askQuestion(rl, "Username",              d.user     || "");
  const portS = await askQuestion(rl, "Port",                  String(d.port || 22));
  d.port     = parseInt(portS, 10) || 22;
  d.identity = await askQuestion(rl, "Identity file",         d.identity || "");
  d.group    = await askQuestion(rl, "Group",                 d.group    || "");
  d.notes    = await askQuestion(rl, "Notes",                 d.notes    || "");

  for (const k of ["identity", "group", "notes"]) {
    if (!d[k]) delete d[k];
  }

  devices[name] = d;
  saveDevices(devices);
  ok(`Updated '${name}'.\n`);
}

async function cmdRemove(devices, name, rl) {
  if (!devices[name]) { err(`No device named '${name}'.`); return; }
  const ans = await askQuestion(rl, `Delete '${name}'? (yes/no)`, "no");
  if (ans.toLowerCase() === "yes" || ans.toLowerCase() === "y") {
    delete devices[name];
    saveDevices(devices);
    ok(`Removed '${name}'.`);
  } else {
    info("Cancelled.");
  }
}

function cmdPing(devices, name) {
  if (!devices[name]) { err(`No device named '${name}'.`); return; }
  const d    = devices[name];
  const host = d.host;
  const port = d.port || 22;
  info(`Checking ${host}:${port} …`);

  return new Promise((resolve) => {
    const socket = new net.Socket();
    socket.setTimeout(5000);
    socket.connect(port, host, () => {
      ok(`${host}:${port} is ${c(GREEN, "reachable")} (SSH port open)\n`);
      socket.destroy();
      resolve();
    });
    socket.on("error", (e) => {
      err(`${host}:${port} is ${c(RED, "unreachable")} — ${e.message}\n`);
      resolve();
    });
    socket.on("timeout", () => {
      err(`${host}:${port} is ${c(RED, "unreachable")} — timeout\n`);
      socket.destroy();
      resolve();
    });
  });
}

async function cmdCopyId(devices, name, rl) {
  if (!devices[name]) { err(`No device named '${name}'.`); return; }
  const d   = devices[name];
  const key = await askQuestion(rl, "Public key path", path.join(os.homedir(), ".ssh", "id_rsa.pub"));
  const args = ["-i", key];
  const port = d.port || 22;
  if (port !== 22) args.push("-p", String(port));
  const userHost = d.user ? `${d.user}@${d.host}` : d.host;
  args.push(userHost);
  info(`Running: ssh-copy-id ${args.join(" ")}`);
  const r = spawnSync("ssh-copy-id", args, { stdio: "inherit", shell: process.platform === "win32" });
  if (r.error && r.error.code === "ENOENT") {
    err("'ssh-copy-id' not found. On Windows, copy your public key manually:");
    info(`cat ~/.ssh/id_rsa.pub | ssh ${userHost} "mkdir -p ~/.ssh && cat >> ~/.ssh/authorized_keys"`);
  }
}

async function cmdExport(devices, rl) {
  const outPath = await askQuestion(rl, "Export to file", "jump_devices.json");
  fs.writeFileSync(outPath, JSON.stringify(devices, null, 2), "utf8");
  ok(`Exported ${Object.keys(devices).length} device(s) to ${outPath}\n`);
}

async function cmdImport(devices, rl) {
  const src = await askQuestion(rl, "Import from file", "jump_devices.json");
  if (!fs.existsSync(src)) { err(`File not found: ${src}`); return; }
  const incoming = JSON.parse(fs.readFileSync(src, "utf8"));
  let merged = 0;
  for (const [name, d] of Object.entries(incoming)) {
    if (devices[name]) {
      const ans = await askQuestion(rl, `'${name}' exists — overwrite? (yes/no)`, "no");
      if (ans.toLowerCase() !== "yes" && ans.toLowerCase() !== "y") continue;
    }
    devices[name] = d;
    merged++;
  }
  saveDevices(devices);
  ok(`Imported ${merged} device(s).\n`);
}

// ── Interactive menu ──────────────────────────────────────────────────────────
async function interactiveMenu() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  // Re-enable raw mode after SSH so readline stays responsive
  rl.on("close", () => process.exit(0));

  let devices = loadDevices();
  banner();

  while (true) {
    divider("saved devices");
    printTable(devices);
    divider("actions");

    const actions = [
      ["c", "connect",  "SSH into a device"],
      ["a", "add",      "Add new device"],
      ["e", "edit",     "Edit a device"],
      ["r", "remove",   "Remove a device"],
      ["p", "ping",     "Ping (check reachability)"],
      ["k", "copy-id",  "Copy SSH key to device"],
      ["x", "export",   "Export devices to JSON"],
      ["i", "import",   "Import devices from JSON"],
      ["q", "quit",     "Exit"],
    ];

    for (const [key, name, desc] of actions) {
      console.log(`  ${c(BOLD + CYAN, `[${key}]`)}  ${c(BOLD, name.padEnd(10))}  ${c(DIM, desc)}`);
    }
    console.log();

    let choice;
    try {
      choice = await new Promise((resolve) => {
        rl.question(`  ${BOLD}>${R} `, resolve);
      });
    } catch {
      break;
    }
    choice = (choice || "").trim().toLowerCase();
    console.log();

    if (choice === "q" || choice === "quit" || choice === "exit") {
      info("Bye!\n");
      rl.close();
      return;
    }

    if (choice === "c" || choice === "connect") {
      const names = Object.keys(devices).sort();
      if (names.length === 0) { warn("No devices. Add one first."); continue; }
      names.forEach((n, i) => console.log(`  ${c(CYAN, `${i+1}.`)} ${n}  ${c(DIM, deviceTag(devices[n]))}`));
      console.log();
      const sel = await askQuestion(rl, "Device name or number");
      if (!sel) continue;
      let target = sel;
      if (/^\d+$/.test(sel)) {
        const idx = parseInt(sel, 10) - 1;
        target = names[idx] || sel;
      }
      if (!devices[target]) { err(`No device '${target}'.`); continue; }
      // Pause readline so SSH gets full tty
      rl.pause();
      doConnect(target, devices[target]);
      rl.resume();
      devices = loadDevices();

    } else if (choice === "a" || choice === "add") {
      await cmdAdd(devices, rl);

    } else if (choice === "e" || choice === "edit") {
      const name = await askQuestion(rl, "Device name to edit");
      if (name) { await cmdEdit(devices, name, rl); devices = loadDevices(); }

    } else if (choice === "r" || choice === "remove") {
      const name = await askQuestion(rl, "Device name to remove");
      if (name) { await cmdRemove(devices, name, rl); devices = loadDevices(); }

    } else if (choice === "p" || choice === "ping") {
      const name = await askQuestion(rl, "Device name to ping");
      if (name) await cmdPing(devices, name);

    } else if (choice === "k" || choice === "copy-id") {
      const name = await askQuestion(rl, "Device name");
      if (name) await cmdCopyId(devices, name, rl);

    } else if (choice === "x" || choice === "export") {
      await cmdExport(devices, rl);

    } else if (choice === "i" || choice === "import") {
      await cmdImport(devices, rl);
      devices = loadDevices();

    } else {
      warn(`Unknown command '${choice}'.`);
    }
  }
}

// ── CLI entry point ───────────────────────────────────────────────────────────
async function main() {
  const args = process.argv.slice(2);
  const cmd  = args[0];
  const arg2 = args[1];

  if (cmd === "--version" || cmd === "-v") {
    console.log(`jump ${VERSION}`);
    return;
  }

  if (cmd === "--help" || cmd === "-h" || cmd === "help") {
    console.log(`
${BOLD}jump${R} ${VERSION} — personal SSH connection manager

${BOLD}USAGE${R}
  jump [command] [name]
  jump                    open interactive menu

${BOLD}COMMANDS${R}
  list                    list all saved devices
  add                     add a new device
  connect  <name>         SSH into a device
  remove   <name>         delete a device
  edit     <name>         edit a device
  ping     <name>         check if SSH port is reachable
  copy-id  <name>         copy your public SSH key to a device
  export                  export devices to a JSON file
  import                  import devices from a JSON file

${BOLD}EXAMPLES${R}
  jump                    (interactive menu)
  jump list
  jump add
  jump connect homeserver
  jump ping nas
`);
    return;
  }

  const devices = loadDevices();

  // Non-interactive commands
  if (cmd === "list") {
    cmdList(devices);
    return;
  }
  if (cmd === "ping" && arg2) {
    await cmdPing(devices, arg2);
    return;
  }
  if (cmd === "connect" && arg2) {
    if (!devices[arg2]) { err(`No device '${arg2}'.`); process.exit(1); }
    doConnect(arg2, devices[arg2]);
    return;
  }

  // Interactive (for add, edit, remove, copy-id, export, import, or no command)
  if (!cmd) {
    await interactiveMenu();
    return;
  }

  // Commands that need readline
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    if (cmd === "add") {
      await cmdAdd(devices, rl);
    } else if (cmd === "edit" && arg2) {
      await cmdEdit(devices, arg2, rl);
    } else if (cmd === "remove" && arg2) {
      await cmdRemove(devices, arg2, rl);
    } else if (cmd === "copy-id" && arg2) {
      await cmdCopyId(devices, arg2, rl);
    } else if (cmd === "export") {
      await cmdExport(devices, rl);
    } else if (cmd === "import") {
      await cmdImport(devices, rl);
    } else {
      err(`Unknown command '${cmd}'. Run  jump --help  for usage.`);
    }
  } finally {
    rl.close();
  }
}

main().catch((e) => {
  if (e.code !== "ERR_USE_AFTER_CLOSE") {
    console.error(e);
    process.exit(1);
  }
});
